### PR TITLE
[2.7] bpo-30759: Copy test_robotparser from master

### PR DESCRIPTION
**Pull request by vstinner**
Copy Lib/test/test_robotparser.py from master to 2.7 and adapt it for
Python 2.7:

* Replace urllib.robotparser with robotparser
* Adjust HTTPServer import
* Replace io.StringIO with StringIO.StringIO
* Remove tests on crawl_delay() and request_rate() since these
  methods were added to Python 3
* Remove subTest()
* Add test_main() which explicitly lists all test cases

### CODE REVIEW

File: Lib/test/test_robotparser.py [hunk 1]
```diff
@@ -1,296 +1,237 @@
-import unittest, StringIO, robotparser
-from test import test_support
-from urllib2 import urlopen, HTTPError
-
-HAVE_HTTPS = True
+import io
```
Comment (serhiy-storchaka) [inline Lib/test/test_robotparser.py]: Is it used?

File: Lib/test/test_robotparser.py [hunk 2]
```diff
@@ -1,296 +1,237 @@
...
+import os
+import unittest
+import robotparser
+from collections import namedtuple
```
Comment (serhiy-storchaka) [inline Lib/test/test_robotparser.py]: Is it used?

**Comment by serhiy-storchaka** [comment]
test_robotparser was rewritten in [bpo-25497](https://bugs.python.org/issue25497) (4da0fd06ce49132d5c2e0525c4a31f807690a9ed) by @berkerpeksag. Maybe try to cherry-pick 4da0fd06ce49132d5c2e0525c4a31f807690a9ed and compare results?

File: Lib/test/test_robotparser.py [hunk 3]
```diff
@@ -1,296 +1,237 @@
...
+        UserAgentWildcardTest,
+        RejectAllRobotsTest,
+        UserAgentOrderingTest,
+        UserAgentGoogleMobileTest,
+        GoogleURLOrderingTest,
+        DisallowQueryStringTest,
+        UseFirstUserAgentWildcardTest,
+        EmptyQueryStringTest,
+        DefaultEntryTest,
+        RobotHandler,
```
Comment (berkerpeksag) [inline Lib/test/test_robotparser.py]: `RobotHandler` is a subclass of `BaseHTTPServer.BaseHTTPRequestHandler`, not `unittest.TestCase`.

**Comment by vstinner** [comment]
Oh, I don't know what happened: I did remove the 2 unused imports, but I see them in the first version of my PR? Well, it doesn't matter, I just removed them :-)

I fixed the 3 reports issues.

Serhiy: "test_robotparser was rewritten in bpo-25497 (4da0fd0) by @berkerpeksag. Maybe try to cherry-pick 4da0fd0 and compare results?"

Sorry, but I don't want to try to *cherry-pick* such large refactoring change from Python 3 to Python 2, the cherry-pick creates a file full of conflcts. IMHO it's simpler to start with a fresh *copy* of the file, and then adapt it for Python 2.

I credited @berkerpeksag in my commit message.

File: Lib/test/test_robotparser.py (hunk 3 above)
Comment (vstinner) [reply Lib/test/test_robotparser.py]: Fixed. Funny that support.run_unittest() doesn't report such bug :-)

File: Lib/test/test_robotparser.py (hunk 2 above)
Comment (vstinner) [reply Lib/test/test_robotparser.py]: removed

File: Lib/test/test_robotparser.py (hunk 1 above)
Comment (vstinner) [reply Lib/test/test_robotparser.py]: removed

**Comment by vstinner** [comment]
@serhiy-storchaka, @berkerpeksag: Does it look better now?

**Review by berkerpeksag — APPROVED** [review]

**Comment by vstinner** [comment]
Ok, thank you for the review @berkerpeksag. Let's go!

**Pull request merged**